### PR TITLE
fix flaky unit test

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR fixes a flaky unit test in the remote process collector tests in workloadmeta component.

### Motivation

The issue was that the mock streaming server was closing the connection right after returning the mock responses, which causes the collector to receive EOF, which calls HandleResync and clears all processes in workloadmeta store, and then re-establishes a new connection.

This is fixed by ensuring the streaming server StreamEntities method blocks indefinitely after sending the responses.

### CODE REVIEW

**Review by gabedos — COMMENTED** [review]
I tried looking for references of StreamEntities on the mockServer but I couldn't find exactly where it's being used. Could you point me to some instances to better understand how it's being used.

**Comment by adel121** [comment]
> I tried looking for references of StreamEntities on the mockServer but I couldn't find exactly where it's being used. Could you point me to some instances to better understand how it's being used.

@gabedos 

This mocks a protobuf service. 
In protobuf, we write `.proto` files with a special syntax to define how the communication content will be serialised and deserialised between the client and the server. You can think of this as a smarter version of JSON and YAML. After writing `.proto` files, we can run a command that generates golang code that allows us to efficiently serialise/deserialise objects based on our definitions. 

The service we are mocking is defined [here](https://github.com/DataDog/datadog-agent/blob/5d1e13226d521281d0e56b5fdbafe7bfa24fd6e1/pkg/proto/datadog/process/workloadmeta_process.proto#L29-L31) with the protobuf syntax:

```
service ProcessEntityStream {
  rpc StreamEntities (ProcessStreamEntitiesRequest) returns (stream ProcessStreamResponse);
}
```

After running the protobuf generate command, [the following interface](https://github.com/DataDog/datadog-agent/blob/5d1e13226d521281d0e56b5fdbafe7bfa24fd6e1/pkg/proto/pbgo/process/workloadmeta_process.pb.go#L870-L873) is generated to represent this service:

```
// ProcessEntityStreamServer is the server API for ProcessEntityStream service.
type ProcessEntityStreamServer interface {
	StreamEntities(*ProcessStreamEntitiesRequest, ProcessEntityStream_StreamEntitiesServer) error
}
```

Protobuf also generates a function that can be used to register an implementation of this service. 

In this unit test, the service is registered [here](https://github.com/DataDog/datadog-agent/blob/0edbbe457e7af0c8fe7bb948a2310d2f93e0b86f/comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go#L266).

So in short, the mockServer you are asking about is implementing the following interface:
```
// ProcessEntityStreamServer is the server API for ProcessEntityStream service.
type ProcessEntityStreamServer interface {
	StreamEntities(*ProcessStreamEntitiesRequest, ProcessEntityStream_StreamEntitiesServer) error
}
```
and it is used to register an implementation of the ProcessEntityStreamServer service. 

Calls to this method can be found in the code generated by protobuf, but you don't need to worry about this, you only need to implement the interface and register the service.

File: comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go [hunk 1]
```diff
@@ -60,7 +60,10 @@ func (s *mockServer) StreamEntities(_ *pbgo.ProcessStreamEntitiesRequest, out pb
 		}
 	}
 
-	return nil
+	// Block indefinitely to avoid closing the stream
+	for {
+		time.Sleep(1 * time.Minute)
+	}
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:66]: I'm wondering if we should introduce a `s.ctx, s.cancelFunc := context.WithCancel(context.Background())` in the mockServer. 

and instead of doing a `for loop`  we do a wait on `ctx.Done()` 

```suggestion
	<-s.ctx.Done()
```

we can add a `func (s *mockServer) Stop() { s.cancelFunc() } 
if in the test we want to properly stop the StreamEntities

Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go]: Good idea 👍 

Updated

File: comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go [hunk 2]
```diff
@@ -254,11 +270,12 @@ func TestCollection(t *testing.T) {
 
 			time.Sleep(time.Second)
 
+			ctx, cancel := context.WithCancel(context.TODO())
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:273]: can I suggested to use `context.Background()` instead of `context.TODO()`

If I remember they are providing similar functionality, however TODO() give the impression this need to be change in the futur. 

```suggestion
			ctx, cancel := context.WithCancel(context.Background())
```

also imo, it doesn't seems that we need a context.WithCancel() at this point because you create another one in the `newMockServer()` function

File: comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go [hunk 3]
```diff
@@ -254,11 +270,12 @@ func TestCollection(t *testing.T) {
...
+
 			// remote process collector server (process agent)
-			server := &mockServer{
-				responses:     test.serverResponses,
-				errorResponse: test.errorResponse,
-			}
+			server := newMockServer(ctx, test.serverResponses, test.errorResponse)
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:276]: based on my previous comment
this we can do 

```suggestion
			server := newMockServer(context.Background(), test.serverResponses, test.errorResponse)
```

File: comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go (hunk 2 above)
Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:273]: Agree about using `context.Background()`

>  at this point because you create another one in the newMockServer() function

You are right, but the created context is used to start the collector [here](https://github.com/DataDog/datadog-agent/blob/79cc32a50cc5e11f788c031cd2f9ce921072b3ef/comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go#L310). I thought I'd use the same context so that if the collector's context is cancelled, the streaming server is stopped automatically.

Comment (clamoriniere) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:273]: Ok didn't catch that you have just moved the ctx creation earlier in the test. 👍 

in that case since the `cancel` function is already all later in the test. maybe you don't need to create another `context.WithCancel()` in the newMockServer() and so no `stop()` function since the cancel will cascade the cancel on your internal `mockserver.ctx`.

Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:273]: Yes this can work too, but in that case we will not be able to stop the streaming server without stopping the collector. 

This is something we might want to test.

Comment (clamoriniere) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go:273]: I feel like the `cancel()` function will always be executed before the `mockServer.stop()` function because we are using a `defer` to call the  `mockServer.stop()` and not for the `cancel()` function

So I suspect that the cancel cascading will always appends before that `mockServer.stop()` is executed.

Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector_test.go]: yes correct
But this is because here we don't need to stop the streaming server before the tests finish.
Calling defer server.stop() is just a safety call to ensure the server is stopped at the end, even if the collector didn't stop.
Basically, with what I have now, there is a dependency as follows:
If the collector stops, the streaming server stops.
We can eliminate this dependency by not cascading the contexts of the server and the collector:

```
ctx := context.Background()

server := newMockServer(ctx)
defer server.stop()

collectorCtx, collectorCancel := context.WithCancel(ctx)

collector.start(collectorCtx)
defer collectorCancel()
```

**Review by clamoriniere — APPROVED** [review]
💯 
thanks for the constructive discussion
